ffmpeg/lpms_ffmpeg.c

```diff
@@ -1021,18 +1049,19 @@ int process_in(struct input_ctx *ictx, AVFrame *frame, AVPacket *pkt)
   // video frames, so continue on to audio.
 
   // Flush video decoder.
-  // To accommodate CUDA, we feed the decoder a a sentinel (flush) frame.
-  // Once the flush frame has been decoded, the decoder is fully flushed.
+  // To accommodate CUDA, we feed the decoder sentinel (flush) frames, till we
+  // get back all sent frames, or we've made SENTINEL_MAX attempts to retrieve
+  // buffered frames with no success.
   // TODO this is unnecessary for SW decoding! SW process should match audio
   if (ictx->vc) {
+    ictx->flushing = 1;
     send_first_pkt(ictx);
-
-    ret = avcodec_receive_frame(ictx->vc, frame);
+    ret = lpms_receive_frame(ictx, ictx->vc, frame);
     pkt->stream_index = ictx->vi;
-    if (!ret) {
-      if (is_flush_frame(frame)) ictx->flushed = 1;
-      return ret;
-    }
+    // Keep flushing if we haven't received all frames back but stop after SENTINEL_MAX tries.
+    if (ictx->pkt_diff != 0 && ictx->sentinel_count <= SENTINEL_MAX) return 0;
```

Comment:
What if `ret` is nonzero? And we're still returning 0 here?

Reply:
Ah, good catch.

So I was returning 0 here because `avcodec_receive_frame` has return value as EAGAIN for our sentinel packets, which we want to ignore for SENTINEL_MAX tries (else we kill the decoder with the error)

I've added a condition here now to only return 0 for EAGAIN and not for some other error.